### ISSUE TITLE
Not working as expected

### ISSUE THREAD

**Issue by Oujiii** [animeloyalty/animesync#31]
Alright, so I'm using a LXC container with Ubuntu 20.04 on my proxmox so I can automate those downloads and I installed `google-chrome-stable` and the other dependencies were installed (they are required for other stuff I was testing). 

I already logged in both Funimation and Crunchyroll over the command line.
This is what happens for Funimation:
```bash
root@anime-dl:~/animesync# animesync download https://www.funimation.com/shows/kaguya-sama-love-is-war
Starting http://localhost:6583/
Sourcing /root/animesync/library
Fetching https://www.funimation.com/shows/kaguya-sama-love-is-war
Skipping https://www.funimation.com/shows/kaguya-sama-love-is-war
```

For Crunchyroll:
```bash
root@anime-dl:~/animesync# animesync download https://www.crunchyroll.com/a-certain-scientific-railgun
Starting http://localhost:6583/
Sourcing /root/animesync/library
Fetching https://www.crunchyroll.com/a-certain-scientific-railgun
Fetching A Certain Scientific Railgun 01 [Crunchyroll]
Finished A Certain Scientific Railgun 01 [Crunchyroll] (00:00:06)
Fetching A Certain Scientific Railgun 02 [Crunchyroll]
Finished A Certain Scientific Railgun 02 [Crunchyroll] (00:00:05)
Fetching A Certain Scientific Railgun 03 [Crunchyroll]

```

I cancelled after the 3rd episode as the episodes just had 0 KB of size. Also, it doesn't work if the account has Crunchyroll Beta, they will launch this new version in less than a month from now and I already noticed that the URL is a little different, so keep that in mind too.

**Comment by Deathspike** [comment]
For *Funimation*, the URL should end with a slash. It's
"https://www.funimation.com/shows/kaguya-sama-love-is-war/"
And not:
"https://www.funimation.com/shows/kaguya-sama-love-is-war".
I tested this, and noticed it throws a different error after that. See #32 

As for *Crunchyroll*, I've tested this on Windows and it's working as intended.
I suspect that there is an issue with ffmpeg; but it should still throw an error if so...
I have no idea what a proxmox is. I'll test on a Ubuntu VM tomorrow (once I spin one up).

**Comment by Oujiii** [comment]
Proxmox is a hypervisor, like HyperV or ESXi. This Ubuntu is headless. Is there a way to run the script with more verbose?

**Comment by Deathspike** [comment]
No. But good that you mention verbose. Animesync does produce logging.
They're in `/root/animesync/logger`. You can check it out and see if you find the root cause.
Or you can attach a file here so I can check what's going on :-)

**Comment by Oujiii** [comment]
```
root@anime-dl:~/animesync# animesync download https://beta.crunchyroll.com/series/GVDHX8DM5/86-EIGHTY-SIX
Starting http://localhost:6583/
Sourcing /root/animesync/library
Fetching https://beta.crunchyroll.com/series/GVDHX8DM5/86-EIGHTY-SIX
Skipping https://beta.crunchyroll.com/series/GVDHX8DM5/86-EIGHTY-SIX
```

```
root@anime-dl:~/animesync/logger# cat 2021-05-23.log 
[2021-05-23T22:23:53.618Z] DEBUG Running animesync (0.11.7)
[2021-05-23T22:23:53.620Z] INFO  Fetching https://beta.crunchyroll.com/series/GVDHX8DM5/86-EIGHTY-SIX
[2021-05-23T22:23:53.631Z] DEBUG HTTP/1.1 GET /api/remote?url=https%3A%2F%2Fbeta.crunchyroll.com%2Fseries%2FGVDHX8DM5%2F86-EIGHTY-SIX
[2021-05-23T22:23:53.641Z] DEBUG HTTP/1.1 200 []
[2021-05-23T22:23:53.648Z] DEBUG HTTP/1.1 GET /api/remote/series?url=https%3A%2F%2Fbeta.crunchyroll.com%2Fseries%2FGVDHX8DM5%2F86-EIGHTY-SIX
[2021-05-23T22:23:53.651Z] DEBUG HTTP/1.1 404 {"statusCode":404,"message":"Error: Not Found\n    at /usr/lib/node_modules/animesync/dist/remote/services/ProviderService.js:70:19\n    at async /usr/lib/node_modules/animesync/dist/remote/services/ProviderService.js:96:20\n    at async /usr/lib/node_modules/animesync/dist/remote/services/PluginService.js:49:20\n    at async PluginService.loadAsync (/usr/lib/node_modules/animesync/dist/remote/services/PluginService.js:57:16)\n    at async PluginService.providersAsync (/usr/lib/node_modules/animesync/dist/remote/services/PluginService.js:46:16)\n    at async ProviderService.providersAsync (/usr/lib/node_modules/animesync/dist/remote/services/ProviderService.js:92:16)\n    at async ProviderService.seriesAsync (/usr/lib/node_modules/animesync/dist/remote/services/ProviderService.js:66:16)\n    at async CacheService.addAsync (/usr/lib/node_modules/animesync/dist/shared/services/CacheService.js:79:27)\n    at async CacheService.getAsync (/usr/lib/node_modules/animesync/dist/shared/services/CacheService.js:68:20)\n    at async RemoteController.seriesAsync (/usr/lib/node_modules/animesync/dist/remote/RemoteController.js:61:16)"}
[2021-05-23T22:23:53.652Z] INFO  Skipping https://beta.crunchyroll.com/series/GVDHX8DM5/86-EIGHTY-SIX
```

This is running beta.

---------------------

The one running normally I will attach as it's too big to paste
[2021.05-23.log](https://github.com/animeloyalty/animesync/files/6529029/2021.05-23.log)

**Comment by Deathspike** [comment]
The beta version of the site isn't supported. I'll add support once the revisions go into production. That's to save myself the effort of rewriting each time while they're in active development, and changing things constantly. I don't want to keep myself too busy on things that are still in flux. For now, avoid using the beta version.

---

The logging is pretty interesting. I see the following:

```
Press [q] to stop, [?] for help
[2021-05-23T22:27:22.286Z] DEBUG frame=    1 fps=0.0 q=-1.0 size=      19kB time=00:00:00.00 bitrate=158720.0kbits/s speed=4.13x
[2021-05-23T22:27:22.286Z] DEBUG [hls @ 0x5da6940] Thread message queue blocking; consider raising the thread_queue_size option (current value: 8)
[2021-05-23T22:27:22.794Z] DEBUG frame=   26 fps=0.0 q=-1.0 size=      19kB time=00:00:05.62 bitrate=  28.2kbits/s speed=  11x
[2021-05-23T22:27:23.302Z] DEBUG frame=   31 fps= 31 q=-1.0 size=      19kB time=00:00:05.62 bitrate=  28.2kbits/s speed=5.53x
[2021-05-23T22:27:23.810Z] DEBUG frame=   36 fps= 24 q=-1.0 size=      19kB time=00:00:05.62 bitrate=  28.2kbits/s speed=3.69x
[2021-05-23T22:27:24.318Z] DEBUG frame=   42 fps= 21 q=-1.0 size=      19kB time=00:00:05.62 bitrate=  28.2kbits/s speed=2.77x
[2021-05-23T22:27:24.608Z] DEBUG [hls @ 0x5da6940] Opening 'crypto+http://localhost:6583/api/rewrite/https%3A%2F%2Fvc.v.vrv.co%2Fevs1%2Fa7a0e426694eb63f47301df17322709d%2Fassets%2Ff5e793af3969e0f96e3c2e14f110d4ac_4034656.mp4%2Fseg-2-v1-a1.ts%3Ft%3Dexp%3D1621981639~acl%3D%2Fevs1%2Fa7a0e426694eb63f47301df17322709d%2Fassets%2Ff5e793af3969e0f96e3c2e14f110d4ac_4034656.mp4%2F*~hmac%3Db71a6aabfce54857bd2bef2c4761a39231e7a50e846ceb95bde0a410e6550846?user-agent=Mozilla%2F5.0%20(X11%3B%20Linux%20x86_64)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Chrome%2F90.0.4430.212%20Safari%2F537.36&origin=https%3A%2F%2Fstatic.crunchyroll.com&referer=https%3A%2F%2Fstatic.crunchyroll.com%2F' for reading
[2021-05-23T22:27:24.618Z] INFO  Finished 86 EIGHTY-SIX 03 [Crunchyroll] (00:00:09)
```

Simply put, *animesync* calls `ffmpeg` to download the video file piped through an emulation proxy. At this point `ffmpeg` starts correctly, and then exits without an error. But it didn't download the whole video. I'll look into this behavior being specific to Linux. Perhaps I've bundled a version of `ffmpeg` that has issues.

**Comment by Oujiii** [comment]
Yeah, that's pretty weird overall. If I'm able to pass the version of ffmpeg I'm using that might help testing too, as I know it works for other tools that are like this one.

**Comment by Deathspike** [comment]
I've added custom `ffmpeg` support:

    animesync settings core --fmpeg /absolute/path/to/ffmpeg

Then `animesync` will launch your executable instead. Hopefully this will assist in tracing this issue on your end.

As for what I've been doing; I now have Ubuntu LTS running with animesync and... it doesn't display the same behavior :-(

**Comment by Oujiii** [comment]
No dice, unfortunately. I know ffmpeg is working properly as there is other tools that I use with it that are working. This is not an Ubuntu VM, but rather an Ubuntu LXC container, what that means is just that it uses the same kernel as the host.

It is just this, forever:
```bash
root@anime-dl:~# tail -f animesync/logger/2021-05-27.log
[2021-05-27T19:44:33.986Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.477x
[2021-05-27T19:44:34.492Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.477x
[2021-05-27T19:44:34.999Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.476x
[2021-05-27T19:44:35.506Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.476x
[2021-05-27T19:44:36.013Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.475x
[2021-05-27T19:44:36.519Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.475x
[2021-05-27T19:44:37.026Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.474x
[2021-05-27T19:44:37.533Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.474x
[2021-05-27T19:44:38.039Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.473x
[2021-05-27T19:44:38.546Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.473x
[2021-05-27T19:44:39.052Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.472x
[2021-05-27T19:44:39.559Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.472x
[2021-05-27T19:44:40.066Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.471x
[2021-05-27T19:44:40.572Z] DEBUG frame= 4915 fps= 10 q=-1.0 size=  213747kB time=00:03:50.00 bitrate=7613.1kbits/s speed=0.471x
```
And the files retains its size of about 209MB

```bash
root@anime-dl:~# ls -lah animesync/library/Ascendance\ of\ a\ Bookworm/
total 209M
drwxr-xr-x 2 root root 4.0K May 27 16:36  .
drwxr-xr-x 3 root root 4.0K May 27 16:36  ..
-rw-r--r-- 1 root root 209M May 27 16:38 'Ascendance of a Bookworm 01 [Crunchyroll].mkv'
```

The only thing I noticed changing is the `speed=0.471x` which is decreasing, I think it will count the download as finished once it reaches absolute 0 (if it does).

**Comment by Deathspike** [comment]
Interesting. The behavior changed after you configured your own `ffmpeg` build. This and #30 leads me to believe that the bundled Linux build is somehow incorrect. I'll look into that.

This new behavior is actually something I have seen in some rare instances. It's like a piece request hangs and never gets cancelled, even though there is code in place to handle that exact situation. But it's a pain in the ass to reproduce on my end, because I rarely get it. But you get it consistently.. 

Since you seem skilled with virtualization, is there any way you could whip up a dockerfile that reproduces the issue? That would be really valuable in tracking down the exact problem and fixing it.

**Comment by Oujiii** [comment]
Unfortunately I am not that skilled in other to create a dockerfile with that issue. 
I just created a new LXC container with a similar config, still not working with the bundled as in #35, but now passing the ffmpeg path it works. The only thing that the first container (the one I opened the issue with) has different from this new one is that it has the iGPU from the server passed through it just because for media streaming, ffmpeg is faster to transcode with it and removes the burden of processing from the CPU.
On the new container it is only using the container and working fine. 
Now I just tested on the original container and it also worked. Reminder that I tested a few times with the old container and it and it always gave that bug that it would stop processing at a random point and it would hang forever. I will test it later to download a few series overnight on both containers and see if they fail or not.

Weirdly enough, the container with the GPU passed is slower than one without it, taking double the time to download one episode (4 min on the original container and 2 min on the new one). I'm not really sure what is happening here. 
Did not try Funimation yet.

**Comment by Deathspike** [comment]
Under Windows, I can spin up a Hyper-V virtual machine or run a container under WSL2. There are no issues running with WSL2, but there are issues when running in Hyper-V. I couldn't make it work reliably using a statically compiled `ffmpeg`. So, I'll remove the bundled `ffmpeg` and rely on the system version instead. That will solve the Linux-specific issues; you just have to bring your own `ffmpeg`.

---

I'll continue this issue report dealing with the 'getting stuck' error. I'll come back onto that topic later.
I'm trying to find a way to reproduce the issue, once I have one, I'll tackle it.

**Comment by Oujiii** [comment]
Thanks. I will keep testing on my end to see if I can still reproduce the issue on my main downloading box. After the latest update (not 0.11.10, didn't test this yet), I noticed that it wasn't getting stuck consistently anymore.

**Comment by Deathspike** [comment]
I've just published a *beta* version (0.11.11-beta.0). Install with:
`npm install -g animesync@beta`
Could you test this version?

**Comment by Deathspike** [comment]
This improvement is now live in `0.11.12`.
